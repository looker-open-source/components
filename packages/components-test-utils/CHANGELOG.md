# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.10.4](https://github.com/looker-open-source/components/compare/v0.10.3...v0.10.4) (2021-01-25)

**Note:** Version bump only for package @looker/components-test-utils





## [0.10.3](https://github.com/looker-open-source/components/compare/v0.10.2...v0.10.3) (2021-01-25)

**Note:** Version bump only for package @looker/components-test-utils





## [0.10.2](https://github.com/looker-open-source/components/compare/v0.10.1...v0.10.2) (2021-01-22)


### Bug Fixes

* On build properly set package.json "files" key ([#1806](https://github.com/looker-open-source/components/issues/1806)) ([c81c4bb](https://github.com/looker-open-source/components/commit/c81c4bb625c58ede49957aad90ac8d9d7b2c4b79))





## [0.10.1](https://github.com/looker-open-source/components/compare/v0.10.0...v0.10.1) (2021-01-22)


### Bug Fixes

* Properly build ems/cjs/ts for easy deep-imports ([#1804](https://github.com/looker-open-source/components/issues/1804)) ([2a523cd](https://github.com/looker-open-source/components/commit/2a523cd70b079944376c6dc87c18202e05a97b01))





# [0.10.0](https://github.com/looker-open-source/components/compare/v0.9.29...v0.10.0) (2021-01-19)


### Bug Fixes

* Properly configure build for tree-shaking support ([9202b72](https://github.com/looker-open-source/components/commit/9202b72bc5fb99eb3e1af7d9e56f3dc15b1df2a4))





## [0.9.15]

### Changed

- `@looker/components-test-utils` helpers will no longer produce `globalStyles`
