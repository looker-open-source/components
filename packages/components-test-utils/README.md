# Test Utilities

Shared functions used for mocking and testing Looker UI Components.
