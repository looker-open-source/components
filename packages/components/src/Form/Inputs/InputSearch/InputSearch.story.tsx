/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import React, { useState, useMemo } from 'react'
import { Story } from '@storybook/react/types-6-0'
import { InputSearch, InputSearchProps } from './InputSearch'

const Template: Story<InputSearchProps> = (args) => <InputSearch {...args} />

export const Basic = Template.bind({})
Basic.args = {}

export const Placeholder = Template.bind({})
Placeholder.args = {
  placeholder: 'Type your search',
}
export const Value = Template.bind({})
Value.args = {
  ...Placeholder.args,
  value: 'Search term',
}

export const Summary = Template.bind({})
Summary.args = {
  ...Placeholder.args,
  summary: '5/10 results',
}

export const DefaultValue = Template.bind({})
DefaultValue.args = {
  ...Placeholder.args,
  defaultValue: 'Default search term',
}

export const NoIcon = Template.bind({})
NoIcon.args = {
  ...Placeholder.args,
  hideSearchIcon: true,
}

export const AutoResize = Template.bind({})
AutoResize.args = {
  autoResize: true,
  maxWidth: 250,
  placeholder: 'Resizes to fit value',
}

export const Disabled = Template.bind({})
Disabled.args = {
  disabled: true,
  value: 'No Search.',
}

export const ReadOnly = Template.bind({})
ReadOnly.args = {
  readOnly: true,
  value: 'Only read',
}
ReadOnly.parameters = {
  storyshots: { disable: true },
}

export const Advanced = () => {
  const [value, setValue] = useState('')

  const options = useMemo(() => {
    const startingOptions = [
      {
        description: 'Really great description',
        detail: 'so cool',
        value: 'Foo',
      },
      { detail: 'got details?', value: 'Bar' },
    ]
    return startingOptions.filter(
      (option) => option.value.toUpperCase().indexOf(value.toUpperCase()) > -1
    )
  }, [value])

  return (
    <InputSearch
      value={value}
      onChange={setValue}
      options={options}
      noOptionsLabel="Nothing matched your search"
      isClearable={false}
      autoFocus
    />
  )
}

export default {
  component: InputSearch,
  title: 'InputSearch',
}
