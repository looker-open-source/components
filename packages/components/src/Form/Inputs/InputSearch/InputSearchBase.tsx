/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import * as CSS from 'csstype'
import omit from 'lodash/omit'
import pick from 'lodash/pick'
import React, {
  FormEvent,
  MouseEvent,
  forwardRef,
  Ref,
  useState,
  useRef,
  ReactElement,
} from 'react'
import styled from 'styled-components'
import { ResponsiveValue, TLengthStyledSystem } from 'styled-system'
import { inputPropKeys } from '../InputProps'
import {
  inputCSS,
  inputHeight,
  InputText,
  InputTextProps,
  inputTextDisabled,
  inputTextFocus,
  inputTextHover,
  inputTextValidation,
} from '../InputText'
import { useControlWarn, useForkedRef, useWrapEvent } from '../../../utils'
import { Flex } from '../../../Layout'

const getHeight = (
  py?: ResponsiveValue<CSS.PaddingProperty<TLengthStyledSystem | symbol>>
) => {
  /* Subtracting vertical padding and border from input text height
  Setting height this way instead of on the parent div allows
  InputChip to expand vertically as needed
  min-height doesn't work because then height: 100% on the children is ignored */
  const verticalSpace =
    typeof py === 'number'
      ? `${((py || 0) + 1) * 2}px`
      : `(${String(py)} * 2) - 2px`
  return `calc(${inputHeight} - ${verticalSpace})`
}

export interface InputSearchBaseProps extends InputTextProps {
  /**
   * hides or customizes search Icon
   */
  searchIcon?: ReactElement | false
  searchControls?: ReactElement
  defaultValue?: string
  value?: string

  onClick?: (e: MouseEvent<HTMLDivElement>) => void
  onMouseDown?: (e: MouseEvent<HTMLDivElement>) => void
  onMouseEnter?: (e: MouseEvent<HTMLDivElement>) => void
  onMouseLeave?: (e: MouseEvent<HTMLDivElement>) => void
  onMouseOver?: (e: MouseEvent<HTMLDivElement>) => void
  onMouseOut?: (e: MouseEvent<HTMLDivElement>) => void
  onMouseUp?: (e: MouseEvent<HTMLDivElement>) => void
}

const InputSearchBaseComponent = forwardRef(
  (
    {
      onChange,
      onClick,
      onMouseDown,
      onMouseEnter,
      onMouseLeave,
      onMouseOut,
      onMouseOver,
      onMouseUp,
      children,
      className,
      defaultValue,
      searchIcon,
      searchControls,
      value: controlledValue = '',
      ...props
    }: InputSearchBaseProps,
    forwardedRef: Ref<HTMLInputElement>
  ) => {
    const isControlled = useControlWarn({
      controllingProps: ['onChange', 'value'],
      isControlledCheck: () => onChange !== undefined,
      name: 'InputSearchBase',
    })
    const [uncontrolledValue, setValue] = useState(
      defaultValue || controlledValue
    )
    const inputValue = isControlled ? controlledValue : uncontrolledValue

    const internalRef = useRef<null | HTMLInputElement>(null)
    const ref = useForkedRef<HTMLInputElement>(internalRef, forwardedRef)

    function handleMouseDown() {
      // set focus to input on mousedown of container
      // need requestAnimationFrame here due to browser updating focus _after_ mousedown is called
      window.requestAnimationFrame(() => {
        internalRef.current && internalRef.current.focus()
      })
    }
    const handleChange = (event: FormEvent<HTMLInputElement>) => {
      setValue(event.currentTarget.value)
      onChange && onChange(event)
    }

    const mouseHandlers = {
      onClick,
      onMouseDown: useWrapEvent(handleMouseDown, onMouseDown),
      onMouseEnter,
      onMouseLeave,
      onMouseOut,
      onMouseOver,
      onMouseUp,
    }

    // 12/17/2019 removing type="search" since React doesn't support onSearch yet
    // resulting in undetectable changes that effect the value
    const input = (
      <InputText
        onChange={handleChange}
        value={inputValue}
        pr="0"
        {...pick(props, inputPropKeys)}
        ref={ref}
      />
    )
    return (
      <div
        className={className}
        {...omit(props, [...inputPropKeys, 'validationType'])}
        {...mouseHandlers}
      >
        {searchIcon && searchIcon}
        {children ? (
          <Flex alignContent="flex-start" flexWrap="wrap">
            {children}
            {input}
          </Flex>
        ) : (
          input
        )}
        {searchControls && searchControls}
      </div>
    )
  }
)

InputSearchBaseComponent.displayName = 'InputSearchBaseComponent'

export const InputSearchBase = styled(InputSearchBaseComponent)`
  ${inputCSS}

  align-items: center;
  display: flex;
  padding: ${({ theme: { space } }) =>
    `${space.xxxsmall} ${space.xxsmall} ${space.xxxsmall}` + ' 0'};
  width: 100%;

  &:hover {
    ${inputTextHover}
  }

  &:focus-within {
    ${inputTextFocus}
  }

  ${(props) => (props.disabled ? inputTextDisabled : '')}

  ${inputTextValidation}

  ${InputText} {
    appearance: none;
    background: transparent;
    border: none;
    box-shadow: none;
    flex: 1;
    height: ${(props) => getHeight(props.py || 2)};

    width: 100%;

    &::-webkit-search-decoration,
    &::-webkit-search-cancel-button,
    &::-webkit-search-results-button,
    &::-webkit-search-results-decoration {
      appearance: none;
    }

    &:focus {
      outline: none;
    }
  }
`
