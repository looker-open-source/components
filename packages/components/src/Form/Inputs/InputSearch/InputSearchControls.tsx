/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */
import { palette } from '@looker/design-tokens'
import omit from 'lodash/omit'
import React, { forwardRef, MouseEvent, Ref } from 'react'
import styled from 'styled-components'
import { Box } from '../../../Layout/Box'
import { IconButton } from '../../../Button'
import { Text } from '../../../Text'

export interface InputSearchControlsProps {
  summary?: string
  showClear: boolean
  onClear: (e: MouseEvent<HTMLButtonElement>) => void
  height?: string
  disabled?: boolean
}

export const InputSearchControlsInternal = forwardRef(
  (
    {
      onClear,
      showClear,
      summary,
      disabled,
      ...props
    }: InputSearchControlsProps,
    ref: Ref<HTMLDivElement>
  ) => {
    if (!summary && !showClear) return null

    const clear = (
      <IconButton
        size="xxsmall"
        icon="Close"
        label="Clear Field"
        onClick={onClear}
        tabIndex={-1}
        disabled={disabled}
      />
    )

    const separator = (
      <Box
        borderRight={`1px solid ${palette.charcoal200}`}
        height="1.5rem"
        style={{
          pointerEvents: 'none',
        }}
      />
    )

    return (
      <div {...omit(props, 'height')} ref={ref}>
        {summary && (
          <Text
            pr="xsmall"
            variant="subdued"
            fontSize="small"
            style={{ whiteSpace: 'nowrap' }}
          >
            {summary}
          </Text>
        )}
        {summary && showClear && separator}
        {showClear && clear}
      </div>
    )
  }
)

InputSearchControlsInternal.displayName = 'InputSearchControlsInternal'

export const InputSearchControls = styled(InputSearchControlsInternal)`
  height: ${props => props.height};
  display: flex;
  align-items: center;
  margin-left: auto;
`
