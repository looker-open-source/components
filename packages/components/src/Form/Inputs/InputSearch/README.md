# InputSearch

The `<InputSearch />` component renders a single search input element.

You can add a placeholder, and a starter value, also a summary text and a closing button associated with it.

```tsx
function Basic(props: InputSearchProps) {
  const { placeholder = 'Type your search', ...restProps } = props;

  return <InputSearch placeholder={placeholder} {...restProps} />;
}
```

## hideSearchIcon

If you don't want to see the search `Icon` use the prop `hideSearchIcon`.

```tsx
function HideSearchIcon(props: InputSearchProps) {
  const {
    hideSearchIcon = false,
    placeholder = 'No search icon here',
    ...restProps
  } = props;

  return (
    <InputSearch
      hideSearchIcon={hideSearchIcon}
      placeholder={placeholder}
      {...restProps}
    />
  );
}
```

## autoResize

Use the `autoResize` prop to allow the input's width to adjust to its current value or placeholder.

```tsx
function AutoResize(props: InputSearchProps) {
  const {
    autoResize = true,
    placeholder = 'Start typing to resize',
    ...restProps
  } = props;

  return (
    <InputSearch
      autoResize={autoResize}
      placeholder={placeholder}
      {...restProps}
    />
  );
}
```

## isClearable

To remove the Clear Field button, use `isClearable={false}`.

```tsx
function IsClearable(props: InputSearchProps) {
  const {
    placeholder = 'Type your search',
    defaultValue = 'test search 1',
    isClearable = false,
    ...restProps
  } = props;

  return (
    <InputSearch
      placeholder={placeholder}
      defaultValue={defaultValue}
      isClearable={isClearable}
      {...restProps}
    />
  );
}
```

## summary

```tsx
function Summary(props: InputSearchProps) {
  const {
    placeholder = 'Type your search',
    summary = 'summary text',
    ...restProps
  } = props;

  return (
    <InputSearch placeholder={placeholder} summary={summary} {...restProps} />
  );
}
```

## defaultValue

Use `defaultValue` when you want to set an initial value but still want InputSearch to control state internally.

```tsx
function DefaultValue(props: InputSearchProps) {
  const {
    placeholder = 'Type your search',
    defaultValue = 'test search 0',
    ...restProps
  } = props;

  return (
    <InputSearch
      placeholder={placeholder}
      defaultValue={defaultValue}
      {...restProps}
    />
  );
}
```

## Controlled Example

```tsx
function Controlled() {
  const [keyword, setKeyword] = useState('Default Value');

  return (
    <InputSearch
      placeholder="Type your search"
      summary={
        keyword.length > 0 ? `You typed ${keyword.length} characters` : ''
      }
      value={keyword}
      onChange={setKeyword}
    />
  );
}
```

## Disabled

Use the `disable` property to make an input field un-editable.

```tsx
function Disable(props: InputSearchProps) {
  const {
    disabled = true,
    placeholder = 'Type your search',
    value = 'Value Disabled',
    ...restProps
  } = props;

  return (
    <SpaceVertical align="start">
      <InputSearch disabled={disabled} />
      <InputSearch disabled={disabled} value={value} />
      <InputSearch
        disabled={disabled}
        placeholder={placeholder}
        {...restProps}
      />
    </SpaceVertical>
  );
}
```

## ReadOnly property

As the name suggests, readOnly makes the text un-editable.

```tsx
function ReadOnly(props: InputSearchProps) {
  const {
    value = "You can't change me.",
    readOnly = true,
    ...restProps
  } = props;

  return <InputSearch value={value} readOnly={readOnly} {...restProps} />;
}
```

## clearIconLabel

The `InputSearch` component provides the prop `clearIconLabel` to customize the input icon label. Hover over the the `x` icon to see the tooltip.

```tsx
function ClearIconLabel(props: InputSearchProps) {
  const {
    clearIconLabel = 'Reset this seach field',
    value = 'my query',
    ...restProps
  } = props;

  return (
    <InputSearch clearIconLabel={clearIconLabel} value={value} {...restProps} />
  );
}
```

## Options (experimental)

The `options` prop offers users suggestions or results as they type, similar to a filterable
[`Select`](/components/forms/select/), but without the intent to restrict the field's value.

Additional props `onSelectOption`, `changeOnSelect` (defaults to true), which determines whether the input
will be updated with the selected option's value, and `clearOnClose` (defaults to the opposite of
`changeOnSelect`), which determines whether the input value will persist after the option list is closed,
(via blur, escape, etc).

```tsx
function Options() {
  const cheeses = [
    {
      description: 'Delicious cheese',
      detail: 'Netherlands ',
      value: 'Gouda',
    },
    { value: 'Cheddar' },
  ];

  const cheeses2 = [{ value: 'Jack' }, { value: 'Swiss' }];

  const handleSelectOption = (option?: SelectOptionObject) =>
    option && alert(`You picked ${option.value}`);

  return (
    <Space>
      <InputSearch
        options={cheeses}
        onSelectOption={handleSelectOption}
        changeOnSelect={false}
        placeholder="Options act like results"
      />
      <InputSearch
        options={cheeses2}
        placeholder="Options act like suggestions"
      />
    </Space>
  );
}
```
