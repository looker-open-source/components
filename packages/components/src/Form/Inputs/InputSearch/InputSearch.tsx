/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import React, {
  forwardRef,
  Ref,
  useEffect,
  useState,
  FormEvent,
  MouseEvent,
} from 'react'
import { InputText, InputTextBaseProps } from '../InputText'
import { InputSearchControls } from './InputSearchControls'

export interface InputSearchProps
  extends Omit<InputTextBaseProps, 'children' | 'value' | 'defaultValue'> {
  summary?: string
  hideControls?: boolean
  hideSearchIcon?: boolean
  onClear?: (e: MouseEvent<HTMLButtonElement>) => void

  defaultValue?: string
  value?: string
}

export const InputSearch = forwardRef(
  (
    {
      summary,
      value: valueProp,
      disabled,
      hideControls = false,
      hideSearchIcon = false,
      onChange,
      onClear,
      defaultValue,
      ...props
    }: InputSearchProps,
    ref: Ref<HTMLInputElement>
  ) => {
    const [value, setValue] = useState('')

    const handleClear = (e: MouseEvent<HTMLButtonElement>) => {
      setValue('')
      onClear && onClear(e)
      onChange &&
        onChange({
          currentTarget: { value: '' },
        } as FormEvent<HTMLInputElement>)
    }

    const handleChange = (e: FormEvent<HTMLInputElement>) => {
      const newValue = (e.target as HTMLInputElement).value
      setValue(newValue)
      onChange && onChange(e)
    }

    // only update when valueProp changes, but not defaultValue
    useEffect(() => {
      setValue(valueProp || defaultValue || '')
      // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [valueProp])

    return (
      <InputText
        {...props}
        value={value}
        ref={ref}
        onChange={handleChange}
        iconBefore={hideSearchIcon ? undefined : 'Search'}
        after={
          !hideControls ? (
            <InputSearchControls
              onClear={handleClear}
              disabled={disabled}
              summary={summary}
              showClear={!!(value && value.length >= 0)}
            />
          ) : undefined
        }
      />
    )
  }
)

InputSearch.displayName = 'InputSearch'
