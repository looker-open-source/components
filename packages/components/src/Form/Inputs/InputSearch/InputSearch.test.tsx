/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import 'jest-styled-components'
import React, { createRef } from 'react'
import { renderWithTheme } from '@looker/components-test-utils'
import { fireEvent, screen } from '@testing-library/react'
import { InputSearch } from './InputSearch'

describe('InputSearch', () => {
  describe('Search icon', () => {
    test('shows by default', () => {
      renderWithTheme(<InputSearch />)
      expect(screen.getByTitle('Search')).toBeInTheDocument()
    })
    test('hidden with hideSearchIcon', () => {
      renderWithTheme(<InputSearch hideSearchIcon />)
      expect(screen.queryByTitle('Search')).not.toBeInTheDocument()
    })
  })

  test('displays placeholder', () => {
    renderWithTheme(<InputSearch placeholder="Type your search" />)
    expect(screen.getByPlaceholderText('Type your search')).toBeVisible()
  })

  test('supports ref assignment', () => {
    const inputRef = createRef<HTMLInputElement>()

    renderWithTheme(<InputSearch ref={inputRef} placeholder="type here" />)
    expect(screen.getByPlaceholderText('type here')).toBe(inputRef.current)
  })

  test('accepts a value', () => {
    renderWithTheme(<InputSearch value="start value" />)
    expect(screen.getByDisplayValue('start value')).toBeVisible()
  })

  test('accepts a defaultValue', () => {
    renderWithTheme(
      <InputSearch defaultValue="replace me" placeholder="type here" />
    )
    const input = screen.getByPlaceholderText('type here')
    expect(input).toHaveValue('replace me')
  })

  test('calls onChange', () => {
    const onChangeMock = jest.fn()
    renderWithTheme(
      <InputSearch onChange={onChangeMock} placeholder="type here" />
    )
    const input = screen.getByPlaceholderText('type here')
    fireEvent.change(input, { target: { value: 'New value' } })
    expect(onChangeMock).toHaveBeenCalledWith('New value')
  })

  test('displays summary', () => {
    renderWithTheme(<InputSearch value="start value" summary="summary value" />)
    expect(screen.getByText('summary value')).toBeVisible()
  })

  describe('Clear button', () => {
    test('clears value', () => {
      renderWithTheme(
        <InputSearch placeholder="type here" defaultValue="start value" />
      )

      const button = screen.getByRole('button')
      fireEvent.click(button)
      expect(screen.getByPlaceholderText('type here')).toHaveDisplayValue('')
      expect(screen.queryByRole('button')).not.toBeInTheDocument()
    })

    test('calls onChange', () => {
      const onChange = jest.fn()

      const { getByRole } = renderWithTheme(
        <InputSearch value="Search" onChange={onChange} />
      )

      const inputButton = getByRole('button')
      inputButton && fireEvent.click(inputButton)
      expect(onChange).toHaveBeenCalledWith('')
    })

    test('clear button and summary together', () => {
      renderWithTheme(
        <InputSearch value="start value" summary="summary value" />
      )
      expect(screen.getByRole('button')).toBeVisible()
      expect(screen.getByText('summary value')).toBeVisible()
    })

    test('hidden when isClearable is false', () => {
      renderWithTheme(
        <InputSearch
          value="start value"
          summary="summary value"
          isClearable={false}
        />
      )
      expect(screen.queryByRole('button')).not.toBeInTheDocument()
    })

    test('hidden when value is empty', () => {
      renderWithTheme(<InputSearch />)
      expect(screen.queryByRole('button')).not.toBeInTheDocument()
    })
  })
})
