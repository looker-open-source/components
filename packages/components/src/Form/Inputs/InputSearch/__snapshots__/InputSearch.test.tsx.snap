// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputSearch default 1`] = `
.c2 {
  width: 20px;
  height: 20px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c2 svg {
  height: 100%;
  width: 100%;
}

.c1 {
  padding-left: 0.25rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #939BA5;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100%;
  pointer-events: none;
}

.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #262D33;
  cursor: text;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: space-evenly;
  -webkit-justify-content: space-evenly;
  -ms-flex-pack: space-evenly;
  justify-content: space-evenly;
  padding: 0.125rem 0.25rem;
  width: 100%;
  height: 36px;
  background: #FFFFFF;
  border: 1px solid #DEE1E5;
  border-radius: 0.25rem;
  color: #343C42;
  font-size: 0.875rem;
}

.c0 .c3 {
  height: 100%;
  max-width: 100%;
  width: 100%;
}

.c0 .c3 span {
  padding: 0 0.5rem;
}

.c0 input {
  background: transparent;
  border: none;
  caret-color: #262D33;
  color: inherit;
  height: 100%;
  outline: none;
  width: 100%;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  font-size: 0.875rem;
  max-width: 100%;
  padding: 0 0.5rem;
}

.c0 input::-webkit-search-decoration,
.c0 input::-webkit-search-cancel-button,
.c0 input::-webkit-search-results-button,
.c0 input::-webkit-search-results-decoration {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c0 input::-webkit-input-placeholder {
  color: #939BA5;
}

.c0 input::-moz-placeholder {
  color: #939BA5;
}

.c0 input:-ms-input-placeholder {
  color: #939BA5;
}

.c0 input::placeholder {
  color: #939BA5;
}

.c0:hover {
  border-color: #C1C6CC;
}

.c0:focus,
.c0:focus-within {
  border-color: #9785F2;
  box-shadow: 0 0 0 2px #E8E5FF;
  outline: none;
}

<div
  className="c0"
  onMouseDown={[Function]}
>
  <div
    className="c1"
  >
    <div
      className="c2"
    >
      <svg
        fill="currentColor"
        height="100%"
        viewBox="0 0 24 24"
        width="100%"
      >
        <title>
          Search
        </title>
        <path
          d="M15.5 14h-.79l-.28-.27A6.471 6.471 0 0016 9.5 6.5 6.5 0 109.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
          fill="currentColor"
        />
      </svg>
    </div>
  </div>
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
</div>
`;

exports[`InputSearch hideSearchIcon removes the icon 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #262D33;
  cursor: text;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: space-evenly;
  -webkit-justify-content: space-evenly;
  -ms-flex-pack: space-evenly;
  justify-content: space-evenly;
  padding: 0.125rem 0.25rem;
  width: 100%;
  height: 36px;
  background: #FFFFFF;
  border: 1px solid #DEE1E5;
  border-radius: 0.25rem;
  color: #343C42;
  font-size: 0.875rem;
}

.c0 .c1 {
  height: 100%;
  max-width: 100%;
  width: 100%;
}

.c0 .c1 span {
  padding: 0 0.5rem;
}

.c0 input {
  background: transparent;
  border: none;
  caret-color: #262D33;
  color: inherit;
  height: 100%;
  outline: none;
  width: 100%;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  font-size: 0.875rem;
  max-width: 100%;
  padding: 0 0.5rem;
}

.c0 input::-webkit-search-decoration,
.c0 input::-webkit-search-cancel-button,
.c0 input::-webkit-search-results-button,
.c0 input::-webkit-search-results-decoration {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c0 input::-webkit-input-placeholder {
  color: #939BA5;
}

.c0 input::-moz-placeholder {
  color: #939BA5;
}

.c0 input:-ms-input-placeholder {
  color: #939BA5;
}

.c0 input::placeholder {
  color: #939BA5;
}

.c0:hover {
  border-color: #C1C6CC;
}

.c0:focus,
.c0:focus-within {
  border-color: #9785F2;
  box-shadow: 0 0 0 2px #E8E5FF;
  outline: none;
}

<div
  className="c0"
  onMouseDown={[Function]}
>
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
</div>
`;
